[CI] Rerun failing tests for periodic build pipelines

breskeby opened the pull request:
When a Buildkite job is retried, the system fetches failed test information from the Develocity API and reruns only the tests that failed in the previous attempt, significantly reducing retry time and CI resource usage.

breskeby committed:
WIP TestRetry

breskeby committed:
Allow jobs in periodic build to retry

breskeby committed:
Rerun only tests failed in previous build job iteration

breskeby committed:
fix spotless

breskeby committed:
Fix typo

breskeby committed:
Make pre command hook more error prone

breskeby committed:
Fix handling of non existing test history

breskeby committed:
Rerun more jobs in period builds

breskeby committed:
Cleanup TestRerun logic

breskeby committed:
Cleanup

breskeby committed:
Fix spotless

breskeby committed:
Apply some further cleanup and polishing

breskeby committed:
Update periodic platform tests

Fix buildkite annotations

Respect retries in buildkite build UIs

Run platform periodic tests with --continue

Add clever retries for windows

Remove dockeravailability

breskeby committed:
Apply review feedback

jozala commented on .buildkite/hooks/pre-command:
```diff
@@ -141,6 +145,103 @@ EOF
+  if BUILD_JSON=$(curl --max-time 30 -H "Authorization: Bearer $BUILDKITE_API_TOKEN" -X GET "https://api.buildkite.com/v2/organizations/elastic/pipelines/${BUILDKITE_PIPELINE_SLUG}/builds/${BUILDKITE_BUILD_NUMBER}?include_retried_jobs=true" 2>/dev/null); then
+    if ORIGIN_JOB_ID=$(printf '%s\n' "$BUILD_JSON" | jq -r --arg jobId "$BUILDKITE_JOB_ID" ' .jobs[] | select(.id == $jobId) | .retry_source.job_id' 2>/dev/null) && [ "$ORIGIN_JOB_ID" != "null" ] && [ -n "$ORIGIN_JOB_ID" ]; then
+      if BUILD_SCAN_URL=$(printf '%s\n' "$BUILD_JSON" | jq -r --arg job_id "$ORIGIN_JOB_ID" '.meta_data["build-scan-" + $job_id]' 2>/dev/null) && [ "$BUILD_SCAN_URL" != "null" ] && [ -n "$BUILD_SCAN_URL" ]; then
+        BUILD_SCAN_ID=$(echo "$BUILD_SCAN_URL" | sed 's|.*/s/||')
```
I think we should start writing some scripts in other language than Bash. That's just a general thought. Nothing to change here I think. I'm just curious about your thoughts after writing this.

jozala commented on .buildkite/hooks/pre-command.bat:
```diff
@@ -17,10 +18,168 @@ set JOB_BRANCH=%BUILDKITE_BRANCH%
+      set SMART_RETRY_STATUS=disabled
+      set SMART_RETRY_DETAILS=
+
+      REM Fetch build information from Buildkite API
```
Ohhh... That's another reason to write these scripts in another language - probably something as OS-agnostic as it can be. 
That's just another general note. Nothing to change here.

jozala commented on build-tools-internal/src/integTest/groovy/org/elasticsearch/gradle/internal/test/rerun/InternalTestRerunPluginFuncTest.groovy:
```diff
@@ -10,279 +10,286 @@
+        $("analysis-common/40_tokenizers/letter")));
+  }
+}
+        '''
```
NIT: Formatting change suggestion:

```suggestion
            import java.util.Formatter;
            
            import org.junit.Test;
            import org.junit.runners.Parameterized;
            
            import com.carrotsearch.randomizedtesting.RandomizedRunner;
            import com.carrotsearch.randomizedtesting.RandomizedTest;
            import com.carrotsearch.randomizedtesting.annotations.Name;
            import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
            import com.carrotsearch.randomizedtesting.annotations.Repeat;
            import com.carrotsearch.randomizedtesting.annotations.Seed;
            import com.carrotsearch.randomizedtesting.annotations.Seeds;
            
            public class SubProject2RandomizedTestClazz2 extends RandomizedTest {
              private String yaml;
            
              public SubProject2RandomizedTestClazz2(
                  @Name("yaml") String yaml) {
                this.yaml = yaml;
              }
            
              @Test
              public void test() {
                System.out.println(yaml + " "
                    + getContext().getRandomness());
              }
            
              @ParametersFactory
              public static Iterable<Object[]> parameters() {
                return Arrays.asList($$(
                    $("analysis-common/30_tokenizers/letter"),
                    $("analysis-common/40_tokenizers/letter")));
              }
            }
            '''.stripMargin()
```

jozala commented on build-tools/src/testFixtures/groovy/org/elasticsearch/gradle/fixtures/AbstractGradleFuncTest.groovy:
```diff
@@ -329,5 +327,76 @@ checkstyle = "com.puppycrawl.tools:checkstyle:10.3"
+
+        File createFailedTest(String clazzName) {
+            createTest(clazzName, testMethodContent(false, true, 1))
+        }
```
Seems to be unused.

jozala reviewed (changes requested):
I can't wait to have this running on CI.
Some minor comments left.
I'd like to discuss two things in `pre-command` about the `--compressed in curl command and `BUILD_SCAN_ID` validation.

breskeby committed:
Simplify BUILD_SCAN_ID resolution in smart retries

- Store build-scan-id directly in Buildkite metadata from Gradle.
- Update pre-command hooks (Bash and Batch) to prefer reading build-scan-id from metadata.
- Remove logic to extract build scan ID from URL in pre-command hooks.

breskeby committed:
Apply review feedback using curl

avoid relying on gunzip and reuse --compressed in curl call

breskeby committed:
More review feedback

- remove unused testfixture method
- use records where possible

breskeby commented:
I triggered a new test run after apply review feedback / minor refactorings https://buildkite.com/elastic/elasticsearch-periodic-platform-support/builds/11538

jozala reviewed (approved):
LGTM

breskeby commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/139430">](https://github.com/elastic/elasticsearch/pull/139430)|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/139431">](https://github.com/elastic/elasticsearch/pull/139431)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/139433">](https://github.com/elastic/elasticsearch/pull/139433)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.8 -->
